File: python/sorting/quick_sort.py
'''
quickSort implementation
'''

def swap(array, i, j):
    temp = array[i]
    array[i] = array[j]
    array[j] = temp

def partition(array, p, r):
    pivot = array[r]
    i = p - 1
    for j in range(p, r):
        if array[j] <= pivot:
            i += 1
            swap(array, i, j)
    swap(array, i+1, r)
    return i + 1

def quickSort(array, p, r):
    if len(array) == 1:
        return array
    if p < r:
        q = partition(array, p, r)
        quickSort(array, p, q-1)
        quickSort(array, q+1, r)

array = [4,2,8,5,3,6,1,7]
quickSort(array, 0, 7)
print(array)

File: go/sorting/quick_sort.go
package main

import (
	"math/rand"
	"fmt"
)

func QuickSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}
	pivot := arr[rand.Intn(len(arr))]

	low := make([]int, 0, len(arr))
	high := make([]int, 0, len(arr))
	mid := make([]int, 0, len(arr))

	for _, item := range arr {
		switch {
			case item < pivot:
				low = append(low, item)	
			case item == pivot:
				mid = append(mid, item)
			case item > pivot:
				high = append(high, item)
		}
	}

	low = QuickSort(low)
	high = QuickSort(high)

	low = append(low, mid...)
	low = append(low, high...)
	
	return low
}

func main() {
	primes := [7]int{11, 17, 5, 7, 2, 13, 3}
	var p []int = primes[:]
	fmt.Println(QuickSort(p))
}